src/posix/platform/trel.cpp

```diff
@@ -0,0 +1,546 @@
+    static char    string[1600];
+
+    uint16_t num = 0;
+    char *   cur = &string[0];
+    char *   end = &string[sizeof(string) - 1];
+
+    cur += snprintf(cur, (uint16_t)(end - cur), "[(len:%d) ", aLength);
+    VerifyOrExit(cur < end);
+
+    while (aLength-- && (num < kMaxWrite))
+    {
+        cur += snprintf(cur, (uint16_t)(end - cur), "%02x ", *aBuffer++);
+        VerifyOrExit(cur < end);
+
+        num++;
+    }
+
+    if (aLength != 0)
+    {
+        cur += snprintf(cur, (uint16_t)(end - cur), "... ");
+        VerifyOrExit(cur < end);
+    }
+
+    *cur++ = ']';
+    VerifyOrExit(cur < end);
+
+    *cur++ = '\0';
+
+exit:
+    *end = '\0';
+    return string;
+}
+
+#endif // #if OPENTHREAD_CONFIG_LOG_PLATFORM && (OPENTHREAD_CONFIG_LOG_LEVEL >= OT_LOG_LEVEL_DEBG)
+
+static void PrepareSocket(uint16_t &aUdpPort)
+{
+    int                 val;
+    struct sockaddr_in6 sockAddr;
+    socklen_t           sockLen;
+
+    otLogDebgPlat("[trel] PrepareSocket()");
+
+    sSocket = SocketWithCloseExec(AF_INET6, SOCK_DGRAM, 0, kSocketNonBlock);
+    VerifyOrDie(sSocket >= 0, OT_EXIT_ERROR_ERRNO);
+
+    // Make the socket non-blocking to allow immediate tx attempt.
+    val = fcntl(sSocket, F_GETFL, 0);
+    VerifyOrDie(val != -1, OT_EXIT_ERROR_ERRNO);
+    val = val | O_NONBLOCK;
+    VerifyOrDie(fcntl(sSocket, F_SETFL, val) == 0, OT_EXIT_ERROR_ERRNO);
+
+    // Bind the socket.
+
+    memset(&sockAddr, 0, sizeof(sockAddr));
+    sockAddr.sin6_family = AF_INET6;
+    sockAddr.sin6_addr   = in6addr_any;
+    sockAddr.sin6_port   = 0;
+
+    if (bind(sSocket, (struct sockaddr *)&sockAddr, sizeof(sockAddr)) == -1)
+    {
+        otLogCritPlat("[trel] Failed to bind socket");
+        DieNow(OT_EXIT_ERROR_ERRNO);
+    }
+
+    sockLen = sizeof(sockAddr);
+
+    if (getsockname(sSocket, (struct sockaddr *)&sockAddr, &sockLen) == -1)
+    {
+        otLogCritPlat("[trel] Failed to get the socket name");
+        DieNow(OT_EXIT_ERROR_ERRNO);
+    }
+
+    aUdpPort = ntohs(sockAddr.sin6_port);
+}
+
+static otError SendPacket(const uint8_t *aBuffer, uint16_t aLength, const otSockAddr *aDestSockAddr)
+{
+    otError             error = OT_ERROR_NONE;
+    struct sockaddr_in6 sockAddr;
+    ssize_t             ret;
+
+    VerifyOrExit(sSocket >= 0, error = OT_ERROR_INVALID_STATE);
+
+    memset(&sockAddr, 0, sizeof(sockAddr));
+    sockAddr.sin6_family = AF_INET6;
+    sockAddr.sin6_port   = htons(aDestSockAddr->mPort);
+    memcpy(&sockAddr.sin6_addr, &aDestSockAddr->mAddress, sizeof(otIp6Address));
+
+    ret = sendto(sSocket, aBuffer, aLength, 0, (struct sockaddr *)&sockAddr, sizeof(sockAddr));
+
+    if (ret != aLength)
+    {
+        otLogDebgPlat("[trel] SendPacket() -- sendto() failed errno %d", errno);
+
+        switch (errno)
+        {
+        case ENETUNREACH:
+        case ENETDOWN:
+        case EHOSTUNREACH:
+            error = OT_ERROR_ABORT;
+            break;
+
+        default:
+            error = OT_ERROR_INVALID_STATE;
```

Comment:
should we return `OT_ERROR_BUSY` for `EAGAIN` and `EWOULDBLOCK`, and `OT_ERROR_FAILED` otherwise?

Reply:
We `INVALID_STATE` to keep the packet in queue and try again. In `SendQueuedPackets()` and `trelSend`.
I prfere not to change this for same reason as https://github.com/openthread/openthread/pull/7125#discussion_r747908930.

Reply:
Fine to me to keep it as is. 

The small issue I see is that it does not treat errors like `EINTR`, `EWOULDBLOCK` and `EAGAIN` differently from other unrecoverable errors.